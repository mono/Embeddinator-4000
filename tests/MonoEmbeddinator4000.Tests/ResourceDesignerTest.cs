﻿using System;
using System.CodeDom.Compiler;
using System.IO;
using IKVM.Reflection;
using Microsoft.CSharp;
using NUnit.Framework;

namespace Embeddinator.Tests
{
    /// <summary>
    /// These are a set of integration/approval tests validating that we are getting expected C# code from ResourceDesignerGenerator
    /// </summary>
    [TestFixture]
    public class ResourceDesignerTest : UniverseTest
    {
        ResourceDesignerGenerator generator;

        [SetUp]
        public override void SetUp()
        {
            base.SetUp();

            generator = new ResourceDesignerGenerator
            {
                OutputDirectory = Environment.CurrentDirectory,
                PackageName = "com.mono.embeddinator",
            };
        }

        Assembly LoadAssembly(string resourceFile)
        {
            var parameters = new CompilerParameters
            {
                OutputAssembly = temp,
            };
            parameters.ReferencedAssemblies.Add(XamarinAndroid.FindAssembly("System.dll"));
            parameters.ReferencedAssemblies.Add(XamarinAndroid.FindAssembly("Mono.Android.dll"));

            AssemblyGenerator.CreateFromResource(resourceFile, parameters);

            foreach (var reference in parameters.ReferencedAssemblies)
            {
                universe.LoadFile(reference);
            }
            return universe.LoadFile(temp);
        }

        void LoadAndGenerate(string csFile)
        {
            var assembly = LoadAssembly(csFile);
            generator.Assemblies = new[] { assembly };
            generator.MainAssembly = assembly.Location;
            generator.Generate();
        }

        [Test]
        public void String()
        {
            LoadAndGenerate("Resource.String");

            string source = generator.ToSource();
            Approvals.Verify(source);
        }

        [Test]
        public void Anim()
        {
            LoadAndGenerate("Resource.Anim");

            string source = generator.ToSource();
            Approvals.Verify(source);
        }

        [Test]
        public void Full()
        {
            LoadAndGenerate("Resource.Full");

            string source = generator.ToSource();
            Approvals.Verify(source);
        }

        [Test]
        public void FullCompile()
        {
            LoadAndGenerate("Resource.Full");

            Assert.IsTrue(generator.WriteAssembly(), "Assembly should compile!");
        }

        [Test]
        public void FullWithResourceFile()
        {
            generator.JavaResourceFile = Path.Combine(Path.GetDirectoryName(GetType().Assembly.Location), "..", "..", "Samples", "R.txt");
            LoadAndGenerate("Resource.Full");

            string source = generator.ToSource();
            Approvals.Verify(source);
        }
    }
}
