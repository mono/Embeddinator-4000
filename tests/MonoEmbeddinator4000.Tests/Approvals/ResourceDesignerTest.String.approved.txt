//------------------------------------------------------------------------------
// <auto-generated>
//     This code was generated by a tool.
//     Runtime Version:4.0.30319.42000
//
//     Changes to this file may cause incorrect behavior and will be lost if
//     the code is regenerated.
// </auto-generated>
//------------------------------------------------------------------------------

[assembly: Android.Runtime.ResourceDesignerAttribute("__embeddinator__.Resource", IsApplication=true)]

namespace @__embeddinator__ {
    using Android.Runtime;
    
    
    [System.CodeDom.Compiler.GeneratedCodeAttribute("Xamarin.Android.Build.Tasks", "1.0.0.0")]
    public class Resource {
        
        private static int ReadFieldInt(System.IntPtr R, string fieldName) {
            System.IntPtr fieldId;
            fieldId = JNIEnv.GetStaticFieldID(R, fieldName, "I");
            return JNIEnv.GetStaticIntField(R, fieldId);
        }
        
        private static int[] ReadFieldArray(System.IntPtr R, string fieldName) {
            System.IntPtr fieldId;
            System.IntPtr value;
            fieldId = JNIEnv.GetStaticFieldID(R, fieldName, "[I");
            value = JNIEnv.GetStaticObjectField(R, fieldId);
            return JNIEnv.GetArray<int>(value);
        }
        
        public static void UpdateIdValues() {
            System.IntPtr R;
            R = JNIEnv.FindClass("com.mono.embeddinator.R$string");
            managedandroid.Resource.String.THIS_IS_CAPS = ReadFieldInt(R, "THIS_IS_CAPS");
            managedandroid.Resource.String.applicationName = ReadFieldInt(R, "applicationName");
            managedandroid.Resource.String.hello = ReadFieldInt(R, "hello");
        }
    }
}
